### PR TITLE
HOSS2-260 - New validation 

### PR DESCRIPTION
**Pull request by burns238**

### CODE REVIEW

File: app/models/DateOfBirth.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import models.ErrorMessage._
+import cats.implicits._
+import java.time.LocalDate
+
+final case class DateOfBirth private (val dob: LocalDate) extends AnyVal
+object DateOfBirth {
+  def apply(dob: LocalDate): ValidationResult[DateOfBirth] =
+    validateDate(dob).map(new DateOfBirth(_))
+
+  def validateDate(dob: LocalDate): ValidationResult[LocalDate] =
```
Comment (liampan) [inline app/models/DateOfBirth.scala:30]: does this need to be public?

File: app/models/DocumentNumber.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+final case class DocumentNumber private (val doc: String) extends AnyVal
+object DocumentNumber {
+  def apply(doc: String): ValidationResult[DocumentNumber] =
+    (
+      validateLength(doc),
+      validateCharacters(doc)
+    ).mapN { case _ => new DocumentNumber(doc) }
+
+  def validateLength(doc: String): ValidationResult[String] =
```
Comment (liampan) [inline app/models/DocumentNumber.scala:32]: ^^

File: app/models/DocumentType.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  case object EUNationalID extends DocumentType("NAT")
+  case object BRC extends DocumentType("BRC")
+  case object BRP extends DocumentType("BRP")
+
+  def apply(docType: String): ValidationResult[DocumentType] = docType match {
+    case "PASSPORT" => Passport.validNec
+    case "NAT"      => EUNationalID.validNec
+    case "BRC"      => BRC.validNec
+    case "BRP"      => BRP.validNec
+    case str        => ErrorMessage("Document type must be PASSPORT, NAT, BRC, or BRP").invalidNec
```
Comment (liampan) [inline app/models/DocumentType.scala:36]: str is not used here

File: app/models/TypeClasses.scala [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  implicit class RichReadsValidationResult[A](reads: Reads[ValidationResult[A]]) {
+    def flattenValidated: Reads[A] = reads.flatMap {
+      case Invalid(errors) =>
+        val validationErrors: List[JsonValidationError] =
+          errors.map(e => JsonValidationError(e.message)).toChain.toList
+        Reads(_ => JsError(Seq(JsPath -> validationErrors)))
+      case Valid(value) => Reads.pure(value)
+    }
+  }
+//   implicit class RichReadsValidationResult[A](reads: Reads[ValidatedNec[ErrorMessage, A]]) {
```
Comment (liampan) [inline app/models/TypeClasses.scala:33]: think this can go

File: app/models/TypeClasses.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+
+package models
+
+import play.api.libs.json.{JsError, JsPath, JsonValidationError, Reads}
+import models.ErrorMessage._
+import cats.data.Validated._
+
+object TypeClasses {
```
Comment (liampan) [inline app/models/TypeClasses.scala:23]: does this want to be a package object, as everything imports it

File: test/controllers/MrzSearchControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      "the validation passes but the connector returns a bad request status" in {
+        tokenCallIsSuccessful
+        val errorResponse =
+          StatusCheckErrorResponse(None, StatusCheckError(ERR_HOME_OFFICE_RESPONSE))
+        val errorResponseWithStatus =
+          StatusCheckErrorResponseWithStatus(BAD_REQUEST, errorResponse)
+        requestMrzCallIsSuccessful(Left(errorResponseWithStatus))
+        await(controller.post(request)) mustEqual withHeaders(
+          BadRequest(Json.toJson(errorResponse)))
+      }
```
Comment (AdamCoops) [inline test/controllers/MrzSearchControllerSpec.scala:96]: minor; lots of unnessesary white space here

File: test/models/NationalitySpec.scala [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      }
+    }
+  }
+
+  "writes" should {
+    "return a JsString" in {
+      Nationality("ABC").map(doc => Json.toJson(doc) shouldEqual JsString("ABC"))
+    }
+  }
+
```
Comment (AdamCoops) [inline test/models/NationalitySpec.scala:71]: WS

File: test/models/DocumentTypeSpec.scala [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        val documentType: DocumentType = DocumentType.BRC
+        Json.toJson(documentType) shouldEqual JsString("BRC")
+      }
+      "type is BRP" in {
+        val documentType: DocumentType = DocumentType.BRP
+        Json.toJson(documentType) shouldEqual JsString("BRP")
+      }
+    }
+  }
+
```
Comment (AdamCoops) [inline test/models/DocumentTypeSpec.scala:90]: WS

File: test/models/DocumentNumberSpec.scala [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      }
+    }
+  }
+
+  "writes" should {
+    "return a JsString" in {
+      DocumentNumber("ABCDE").map(doc => Json.toJson(doc) shouldEqual JsString("ABCDE"))
+    }
+  }
+
```
Comment (AdamCoops) [inline test/models/DocumentNumberSpec.scala:94]: WS

File: test/models/DateOfBirthSpec.scala [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    }
+  }
+
+  "writes" should {
+    "return a JsString" in {
+      DateOfBirth(LocalDate.now.minusDays(1)).map(doc =>
+        Json.toJson(doc) shouldEqual JsString(LocalDate.now.minusDays(1).toString))
+    }
+  }
+
```
Comment (AdamCoops) [inline test/models/DateOfBirthSpec.scala:69]: WS

File: test/controllers/NinoSearchControllerSpec.scala [hunk 11]
```diff
@@ -16,31 +16,26 @@
...
 package controllers
 
-import org.mockito.ArgumentMatchers.any
-import org.mockito.Mockito.when
-import scala.concurrent.Future
-import scala.concurrent.ExecutionContext.Implicits.global
 import play.api.libs.json.{JsValue, Json}
 import play.api.test.FakeRequest
-import models.{OAuthToken, StatusCheckByNinoRequest, StatusCheckError, StatusCheckErrorResponse, StatusCheckErrorResponseWithStatus, StatusCheckResponse, StatusCheckResult}
+import models._
```
Comment (AdamCoops) [inline test/controllers/NinoSearchControllerSpec.scala:21]: 👏

File: it/uk/gov/hmrc/homeofficeimmigrationstatusproxy/connectors/HomeOfficeRightToPublicFundsConnectorISpec.scala [hunk 12]
```diff
@@ -137,4 +148,100 @@ class HomeOfficeRightToPublicFundsConnectorISpec
...
+    "return unknown error if 5xx response" in {
+      givenStatusPublicFundsByMrzStub(INTERNAL_SERVER_ERROR, validMrzRequestBody, "")
+
+      val result: Either[StatusCheckErrorResponseWithStatus, StatusCheckResponse] =
+        connector.statusPublicFundsByMrz(mrzRequest, dummyCorrelationId, dummyOAuthToken).futureValue
+
+      result.left.get.statusCode shouldBe INTERNAL_SERVER_ERROR
+      result.left.get.errorResponse.error.errCode shouldBe ERR_UNKNOWN
+    }
+
```
Comment (AdamCoops) [inline it/uk/gov/hmrc/homeofficeimmigrationstatusproxy/connectors/HomeOfficeRightToPublicFundsConnectorISpec.scala:243]: WS

File: it/uk/gov/hmrc/homeofficeimmigrationstatusproxy/controllers/MrzSearchControllerISpec.scala [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    extends ServerBaseISpec with HomeOfficeRightToPublicFundsStubs with JsonMatchers {
+  this: Suite with ServerProvider =>
+
+  val url = s"http://localhost:$port"
+
+  lazy val wsClient: WSClient = app.injector.instanceOf[WSClient]
+
+  def ping: WSResponse = wsClient.url(s"$url/ping/ping").get.futureValue
+
+  def publicFundsByMrz(payload: String, correlationId: String = "sjdfhks123"): WSResponse =
```
Comment (liampan) [inline it/uk/gov/hmrc/homeofficeimmigrationstatusproxy/controllers/MrzSearchControllerISpec.scala:23]: sjdfhks123 can we make this something more meaningful?

**Review by liampan — APPROVED** [review]

**Pull request merged**